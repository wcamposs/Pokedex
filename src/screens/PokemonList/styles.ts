// libraries
import { StyleSheet } from 'react-native';

const styles = StyleSheet.create({
    container: {
        flex: 1,
    },
    searchbarContainer: {
        marginTop: -16,
        marginBottom: 12,
    },
    loadingContainer: {
        alignItems: 'center',
        justifyContent: 'center',
        marginBottom: 10,
    },
    listContainer: {
        paddingVertical: 8,
    },
});

export default styles;
