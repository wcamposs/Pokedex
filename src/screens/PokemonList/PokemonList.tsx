// libraries
import React, { useEffect, useRef, useState } from 'react';
import {
	ActivityIndicator,
	FlatList,
	Keyboard,
	SafeAreaView,
	View,
} from 'react-native';

// js
import styles from './styles';

// services
import PokemonService from '../../services/PokemonService';

// utils
import colors from '../../utils/colors';

// interfaces
import { Pokemon } from '../../utils/interfaces';

// components
import Header from '../../components/Header/Header';
import EmptyList from '../../components/EmptyList/EmptyList';
import SearchBar from '../../components/SearchBar/SearchBar';
import PokemonCard from '../../components/PokemonCard/PokemonCard';

function PokemonList({ navigation }: any) {
	// states
	const [loading, setLoading] = useState(false);
	const [pokemonList, setPokemonList] = useState([]);
	const [searchParams, setSearchParams] = useState('');
	const [shouldRenderEmpty, setShouldRenderEmpty] = useState(false);

	const typingTimeout = useRef(0);

	useEffect(() => {
		if (typingTimeout.current) {
			clearTimeout(typingTimeout.current);
		}

		typingTimeout.current = window.setTimeout(
			() => handleLoadPokemons(),
			750,
		);
	}, [searchParams]);

	// methods
	function handleLoadPokemons() {
		if (searchParams.length) {
			PokemonService.loadSearchedPokemons(
				searchParams,
				setLoading,
				setShouldRenderEmpty,
				setPokemonList,
			);
		} else {
			PokemonService.loadPokemons(
				setLoading,
				pokemonList,
				setShouldRenderEmpty,
				setPokemonList,
			);
		}
	}

	function loadMorePokemons() {
		if (!searchParams.length) {
			PokemonService.loadPokemons(
				setLoading,
				pokemonList,
				setShouldRenderEmpty,
				setPokemonList,
			);
		}
	}

	function onBlur() {
		Keyboard.dismiss();
	}

	function onChangeText(text: string = '') {
		if (!text.length) {
			setPokemonList([]);
		}

		setSearchParams(text);
	}

	// renders
	function renderPokemonCard({ item }: { item: Pokemon }) {
		return (
			<PokemonCard
				key={item.name}
				pokemon={item}
				navigation={navigation}
			/>
		);
	}

	function renderEmptycomponent() {
		if (!shouldRenderEmpty || loading) {
			return null;
		}

		return <EmptyList />;
	}

	function renderFooterComponent() {
		return (
			<>
				{!loading ? null : (
					<View style={styles.loadingContainer}>
						<ActivityIndicator
							style={styles.loading}
							size={70}
							color={colors.red}
						/>
					</View>
				)}
			</>
		);
	}

	return (
		<SafeAreaView style={styles.container}>
			<Header navigation={navigation} title="Pokedex" />
			<View style={styles.searchbarContainer}>
				<SearchBar
					onBlur={onBlur}
					onChangeText={onChangeText}
					value={searchParams}
				/>
			</View>
			<FlatList
				style={styles.listContainer}
				maxToRenderPerBatch={5}
				onEndReached={loadMorePokemons}
				onEndReachedThreshold={0.5}
				data={pokemonList}
				renderItem={renderPokemonCard}
				keyExtractor={(_, index) => 'item_' + index}
				ListFooterComponent={renderFooterComponent}
				ListEmptyComponent={renderEmptycomponent}
				showsVerticalScrollIndicator={false}
			/>
		</SafeAreaView>
	);
}

export default PokemonList;
