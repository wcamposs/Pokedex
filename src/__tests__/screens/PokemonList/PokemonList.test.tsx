import React from 'react';
import renderer from 'react-test-renderer';

import PokemonList from '../../../screens/PokemonList/PokemonList';

jest.useFakeTimers();

describe('<Pokemon List />', () => {
	it('renders correctly', () => {
		const navigation = jest.fn();

		const tree = renderer
			.create(<PokemonList navigation={navigation} />)
			.toJSON();
		expect(tree).toMatchSnapshot();
	});
});
