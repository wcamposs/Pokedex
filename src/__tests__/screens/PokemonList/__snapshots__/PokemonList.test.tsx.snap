// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pokemon List /> renders correctly 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "transparent",
        "marginTop": 20,
        "padding": 24,
        "zIndex": 100,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "space-between",
        }
      }
    >
      <View />
      <View
        style={
          Object {
            "alignItems": "center",
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#d40000",
              "fontSize": 32,
              "fontWeight": "bold",
              "letterSpacing": 1,
              "lineHeight": 32,
              "paddingTop": 12,
              "textAlign": "center",
              "textTransform": "uppercase",
            }
          }
        >
          Pokedex
        </Text>
      </View>
      <View />
    </View>
  </View>
  <View
    style={
      Object {
        "marginBottom": 12,
        "marginTop": -16,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#FFF",
          "borderColor": "#e6e6f0",
          "borderRadius": 12,
          "borderWidth": 0.5,
          "flexDirection": "row",
          "height": 42,
          "marginHorizontal": 12,
          "marginTop": 12,
          "paddingHorizontal": 12,
        }
      }
    >
      <TextInput
        allowFontScaling={true}
        autoCapitalize="none"
        autoCompleteType="off"
        onBlur={[Function]}
        onChangeText={[Function]}
        placeholder="Search a pokemon..."
        rejectResponderTermination={true}
        returnKeyType="search"
        selectTextOnFocus={true}
        style={
          Object {
            "flex": 1,
            "paddingHorizontal": 12,
            "paddingVertical": 12,
          }
        }
        underlineColorAndroid="transparent"
        value=""
      />
    </View>
  </View>
  <RCTScrollView
    ListEmptyComponent={[Function]}
    ListFooterComponent={[Function]}
    data={Array []}
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={5}
    onContentSizeChange={[Function]}
    onEndReached={[Function]}
    onEndReachedThreshold={0.5}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    showsVerticalScrollIndicator={false}
    stickyHeaderIndices={Array []}
    style={Object {}}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
      />
    </View>
  </RCTScrollView>
</RCTSafeAreaView>
`;
